securedrop/passphrases.py

```diff
@@ -0,0 +1,138 @@
+from pathlib import Path
+from random import SystemRandom
+from typing import Optional, NewType, Set
+
+from typing import Dict
+from typing import List
+
+from sdconfig import config
+
+# A list of words to be used by as a passphrase
+# For example: "recede anytime acorn durably discuss"
+# More details at https://www.rempe.us/diceware/#eff
+DicewarePassphrase = NewType("DicewarePassphrase", str)
+
+
+_current_generator = None  # type: Optional["PassphraseGenerator"]
+
+
+class InvalidWordListError(Exception):
+    pass
+
+
+class PassphraseGenerator:
+
+    PASSPHRASE_WORDS_COUNT = 7
+
+    # Enforce a reasonable maximum length for passphrases to avoid DoS
+    MAX_PASSPHRASE_LENGTH = 128
+    MIN_PASSPHRASE_LENGTH = 20
+
+    _WORD_LIST_MINIMUM_SIZE = 6000  # Minimum number of words in any of the word lists
```

Comment:
Is 6000 enough?

Reply:
Good question. 

According to [the Diceware FAQ](https://theworld.com/%7Ereinhold/dicewarefaq.html), Diceware's entropy is calculated as ~12.92 bits per word, based on a list of 7776 (6<sup>5</sup>) words. The formula is <code>M * log2(N)</code>, where `M` is the number of words and `N` the length of the word list, so a seven-word passphrase from 7776 words provides 90.4 bits of entropy. Our current English list should produce 90.25, the French 89.95.

The length of the words is less important then their number, as each is effectively a symbol in a large alphabet. In fact the originator of Diceware intentionally included short, meaningless words to reduce the length of passphrases and make them easier to remember and type. EFF produced Diceware lists with words shorter than three eliminated, but in their own words, that was for usability: ["Note that the security of a passphrase generated using either list is identical; the differences are in usability, including memorability, not in security."](https://www.eff.org/deeplinks/2016/07/new-wordlists-random-passphrases)

So I don't think we should just eliminate short words. It's counter-intuitive, but that would actually reduce the entropy of our passphrases, to 89.41 bits for English at 7000 words, or 88.34 for French at 6300.

My take is that we should definitely have a lower limit on the length of any word list we use for codenames, but it should probably be 7500 or more, to ensure 90 bits of entropy. Perhaps the best way forward is to replace our current English list with the EFF list, not because the short words are a security problem, but because I think the EFF list will produce better machine translations to our other supported languages.

For the purposes of this PR, I think we should set `_WORD_LIST_MINIMUM_SIZE` to 7300 and eliminate the word length requirement when loading word lists. That will keep French codenames working, while ensuring 89.83 bits of entropy.